### PR TITLE
libazureinit: rustdocs for error, http, goalstate, IMDS, unittest

### PR DESCRIPTION
**Pull request by dongsupark**
Add rustdocs for error, http, goalstate, IMDS, unittest with examples if possible.

See also https://github.com/Azure/azure-init/issues/127.

## Testing done

Rustdocs can be checked via web like:
```
$ cargo doc --open
```

Rustdocs tests passed.

```
$ cargo test --doc
```

### CODE REVIEW

**Review by jeremycline — CHANGES_REQUESTED** [review]
Awesome, thanks for adding all this documentation. I have a few little nit-picks, but nothing major.

File: libazureinit/src/error.rs [hunk 1]
```diff
@@ -1,6 +1,20 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+/// Set of error codes that can be used libazureinit as well as its relevant code.
```
[the remark this answers is not in the document]
Comment (dongsupark) [reply libazureinit/src/error.rs]: Done

File: libazureinit/src/goalstate.rs [hunk 2]
```diff
@@ -25,6 +50,7 @@ pub struct Goalstate {
     incarnation: String,
 }
 
+/// Container of goalstate of the virtual machine. Metadata is written in XML format.
```
[the remark this answers is not in the document]
Comment (dongsupark) [reply libazureinit/src/goalstate.rs]: Thanks for the tip, done

File: libazureinit/src/goalstate.rs [hunk 3]
```diff
@@ -48,6 +76,31 @@ pub struct RoleInstance {
...
+/// variable with possible options, to pass it as parameter.
+///
+/// Parameter url is optional. If None is passed, it defaults to
+/// DEFAULT_GOALSTATE_URL, an internal goalstate URL available in the Azure VM.
+///
+/// # Example
+///
+/// ```
+/// # use reqwest::Client;
+/// # use std::time::Duration;
```
[the remark this answers is not in the document]
Comment (dongsupark) [reply libazureinit/src/goalstate.rs]: Done

File: libazureinit/src/error.rs [hunk 4]
```diff
@@ -1,6 +1,20 @@
...
+/// # Example
+///
+/// ```rust
+/// # use libazureinit::error::Error;
+///
+/// fn do_someuser() -> Result<(), Error> {
+///     return Err(Error::UserMissing {
+///         user: "testuser1".to_string(),
+///     });
+/// }
```
[the remark this answers is not in the document]
Comment (dongsupark) [reply libazureinit/src/error.rs:15]: Agreed in general, but I am having hard time to write a clean compilable example code for that. Will look into it tomorrow again.

[the remark this answers is not in the document]
Comment (dongsupark) [reply libazureinit/src/error.rs]: Updated the example code, thanks.

**Pull request merged**